extensions/scarb-doc/src/types.rs

```diff
@@ -304,15 +529,83 @@ pub struct Impl {
     pub id: ImplDefId,
     pub node: ast::ItemImplPtr,
 
+    pub impl_types: Vec<ImplType>,
+    pub impl_functions: Vec<ImplFunction>,
+
     pub item_data: ItemData,
+    pub lookup_item_data: LookupItemData,
 }
 
 impl Impl {
-    pub fn new(db: &dyn DefsGroup, id: ImplDefId, node: &ast::ItemImpl) -> Self {
+    pub fn new(db: &RootDatabase, id: ImplDefId) -> Self {
+        let impl_types = db.impl_types(id).unwrap();
+        let impl_types = impl_types
+            .iter()
+            .map(|(id, _)| ImplType::new(db, *id))
+            .collect::<Vec<_>>();
+
+        let impl_functions = db.impl_functions(id).unwrap();
+        let impl_functions = impl_functions
+            .iter()
+            .map(|(_name, id)| ImplFunction::new(db, *id))
+            .collect::<Vec<_>>();
+
+        let node = id.stable_ptr(db);
+        Self {
+            id,
+            node,
+            impl_types,
+            impl_functions,
+            item_data: ItemData::new(db, id, node),
+            lookup_item_data: LookupItemData::new(
+                db,
+                LookupItemId::ModuleItem(ModuleItemId::Impl(id)),
+            ),
+        }
+    }
+}
+
+#[derive(Clone, Debug)]
+pub struct ImplType {
+    pub id: ImplTypeDefId,
+    pub node: ast::ItemTypeAliasPtr,
+
+    pub item_data: ItemData,
+    pub lookup_item_data: LookupItemData,
+}
+
+impl ImplType {
+    pub fn new(db: &RootDatabase, id: ImplTypeDefId) -> Self {
+        let node = id.stable_ptr(db);
+        Self {
+            id,
+            node,
+            item_data: ItemData::new(db, id, node),
+            lookup_item_data: LookupItemData::new(db, LookupItemId::ImplItem(ImplItemId::Type(id))),
+        }
+    }
+}
+
+#[derive(Clone, Debug)]
+pub struct ImplFunction {
+    pub id: ImplFunctionId,
+    pub node: ast::FunctionWithBodyPtr,
+
+    pub item_data: ItemData,
+    pub lookup_item_data: LookupItemData,
+}
```

Comment:
This also seems like something that could be generalised by a function type

Reply:
Depends if we want to keep it isomorphic to structs from compiler or not

Reply:
What do you think?

Reply:
I would keep it this way for now, if we don't find any use case for that we can merge it